admin/cli/cli.go

```diff
@@ -0,0 +1,193 @@
+// pmm-admin
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package cli
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+
+	"github.com/alecthomas/kong"
+	"github.com/sirupsen/logrus"
+
+	"github.com/percona/pmm/admin/commands"
+	"github.com/percona/pmm/admin/commands/inventory"
+	"github.com/percona/pmm/admin/commands/management"
+	"github.com/percona/pmm/version"
+)
+
+var isJSON = false
+
+type CLIGlobalFlags struct {
+	ServerURL               string      `placeholder:"SERVER-URL" help:"PMM Server URL in https://username:password@pmm-server-host/ format"`
+	SkipTLSCertificateCheck bool        `name:"server-insecure-tls" help:"Skip PMM Server TLS certificate validation"`
+	EnableDebug             bool        `name:"debug" help:"Enable debug logging"`
+	EnableTrace             bool        `name:"trace" help:"Enable trace logging (implies debug)"`
+	PMMAgentListenPort      uint32      `default:"${defaultListenPort}" help:"Set listen port of pmm-agent"`
+	JSON                    jsonFlag    `help:"Enable JSON output"`
+	Version                 versionFlag `short:"v" help:"Show application version"`
+}
+
+type versionFlag bool
+
+func (v versionFlag) BeforeApply(app *kong.Kong, ctx *kong.Context) error {
```

Comment:
What if we make it `AfterApply` instead of `BeforeApply`? can we get access to json flag without hacks?

Reply:
And probably we can get a value of flag via `https://github.com/alecthomas/kong/blob/v0.6.1/context.go#L127`

Reply:
in that case we won't need global variable `isJSON`

Reply:
With `AfterApply` kong already prints an error message a command is missing and prints help. This is not usable unfortunately.

Since kong processes flags left to right, I could not find a better way around this.
I considered a "default" command which would print either help or version but then the help's context is the command and not the whole cli app so the help does not include the information we expect.